contracts/src/v0.8/functions/dev/1_0_0/FunctionsSubscriptions.sol

```diff
@@ -0,0 +1,551 @@
+  error Reentrant();
+
+  LinkTokenInterface private LINK;
+
+  // ================================================================
+  // |                      Subscription state                      |
+  // ================================================================
+
+  // We make the sub count public so that its possible to
+  // get all the current subscriptions via getSubscription.
+  uint64 private s_currentsubscriptionId;
+
+  // s_totalBalance tracks the total LINK sent to/from
+  // this contract through onTokenTransfer, cancelSubscription and oracleWithdraw.
+  // A discrepancy with this contract's LINK balance indicates that someone
+  // sent tokens using transfer and so we may need to use recoverFunds.
+  uint96 private s_totalBalance;
+
+  mapping(uint64 => IFunctionsSubscriptions.Subscription) /* subscriptionId */ /* subscription */
+    private s_subscriptions;
+
+  // We need to maintain a list of addresses that can consume a subscription.
+  // This bound ensures we are able to loop over them as needed.
+  // Should a user require more consumers, they can use multiple subscriptions.
+  uint16 public constant MAX_CONSUMERS = 100;
+  mapping(address => mapping(uint64 => IFunctionsSubscriptions.Consumer)) /* consumer */ /* subscriptionId */ /* Consumer data */
+    private s_consumers;
+
+  event SubscriptionCreated(uint64 indexed subscriptionId, address owner);
+  event SubscriptionFunded(uint64 indexed subscriptionId, uint256 oldBalance, uint256 newBalance);
+  event SubscriptionConsumerAdded(uint64 indexed subscriptionId, address consumer);
+  event SubscriptionConsumerRemoved(uint64 indexed subscriptionId, address consumer);
+  event SubscriptionCanceled(uint64 indexed subscriptionId, address fundsRecipient, uint256 fundsAmount);
+  event SubscriptionOwnerTransferRequested(uint64 indexed subscriptionId, address from, address to);
+  event SubscriptionOwnerTransferred(uint64 indexed subscriptionId, address from, address to);
+
+  error TooManyConsumers();
+  error InsufficientSubscriptionBalance();
+  error InvalidConsumer(uint64 subscriptionId, address consumer);
+  error ConsumerRequestsInFlight();
+  error InvalidSubscription();
+  error OnlyCallableFromLink();
+  error InvalidCalldata();
+  error MustBeSubOwner(address owner);
+  error PendingRequestExists();
+  error MustBeRequestedOwner(address proposedOwner);
+  error BalanceInvariantViolated(uint256 internalBalance, uint256 externalBalance); // Should never happen
+  event FundsRecovered(address to, uint256 amount);
+
+  // @dev NOP balances are held as a single amount. The breakdown is held by the Coordinator.
+  mapping(address => uint96) /* Coordinator => LINK balance (Juels) */ private s_withdrawableTokens;
+
+  // ================================================================
+  // |                       Request state                          |
+  // ================================================================
+
+  struct Request {
```

Comment:
Shall this include don ID?

Reply:
If it includes the coordinator address, then do we need the DON ID?